js/modules/k6/experimental/sse/sse.go

```diff
@@ -0,0 +1,540 @@
+type Client struct {
+	rt            *goja.Runtime
+	ctx           context.Context
+	url           string
+	resp          *http.Response
+	eventHandlers map[string][]goja.Callable
+	done          chan struct{}
+	shutdownOnce  sync.Once
+
+	tagsAndMeta    *metrics.TagsAndMeta
+	samplesOutput  chan<- metrics.SampleContainer
+	builtinMetrics *metrics.BuiltinMetrics
+}
+
+// HTTPResponse is the http response returned by sse.open.
+type HTTPResponse struct {
+	URL     string            `json:"url"`
+	Status  int               `json:"status"`
+	Headers map[string]string `json:"headers"`
+	Error   string            `json:"error"`
+}
+
+// Event represents a Server-Sent Event
+type Event struct {
+	ID   string
+	Name string
+	Data string
+}
+
+type sseOpenArgs struct {
+	setupFn     goja.Callable
+	headers     http.Header
+	method      string
+	body        string
+	cookieJar   *cookiejar.Jar
+	tagsAndMeta *metrics.TagsAndMeta
+}
+
+// Exports returns the exports of the sse module.
+func (mi *sse) Exports() modules.Exports {
+	return modules.Exports{Default: mi.obj}
+}
+
+// Open establishes a http client connection based on the parameters provided.
+func (mi *sse) Open(url string, args ...goja.Value) (*HTTPResponse, error) {
+	ctx := mi.vu.Context()
+	rt := mi.vu.Runtime()
+	state := mi.vu.State()
+	if state == nil {
+		return nil, ErrSSEInInitContext
+	}
+
+	parsedArgs, err := parseConnectArgs(state, rt, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedArgs.tagsAndMeta.SetSystemTagOrMetaIfEnabled(state.Options.SystemTags, metrics.TagURL, url)
+
+	client, connEndHook, err := mi.open(ctx, state, rt, url, parsedArgs)
+	defer connEndHook()
+	if err != nil {
+		// Pass the error to the user script before exiting immediately
+		client.handleEvent("error", rt.ToValue(err))
+		if state.Options.Throw.Bool {
+			return nil, err
+		}
+		return client.wrapHTTPResponse(err.Error())
+	}
+
+	// Run the user-provided set up function
+	if _, err := parsedArgs.setupFn(goja.Undefined(), rt.ToValue(&client)); err != nil {
+		_ = client.closeResponseBody()
+		return nil, err
+	}
+
+	// The connection is now open, emit the event
+	client.handleEvent("open")
+
+	readEventChan := make(chan Event)
+	readErrChan := make(chan error)
+	readCloseChan := make(chan int)
+
+	// Wraps a couple of channels
+	go client.readEvents(readEventChan, readErrChan, readCloseChan)
+
+	// This is the main control loop. All JS code (including error handlers)
+	// should only be executed by this thread to avoid race conditions
+	for {
+		select {
+		case event := <-readEventChan:
+			metrics.PushIfNotDone(ctx, client.samplesOutput, metrics.Sample{
+				TimeSeries: metrics.TimeSeries{
+					Metric: client.builtinMetrics.SSEEventReceived,
+					Tags:   client.tagsAndMeta.Tags,
+				},
+				Time:     time.Now(),
+				Metadata: client.tagsAndMeta.Metadata,
+				Value:    1,
+			})
+
+			client.handleEvent("event", rt.ToValue(event))
+
+		case readErr := <-readErrChan:
```

Comment:
Errors should also be measured with metrics, as in the http module. Instead of handling errors at the JavaScript level, it is much more convenient to set a threshold for the error metric...

Reply:
I kind of disagree on this - we do more or less the same for websockets. Especially as a lot of the cases it is just an error to close. 

This still allows users to create custom metrics for errors, but doesn't force them.

We do have error metrics when teh connection couldn't be started though.

Also not SSE expert - so maybe not as similar :shrug:

Reply:
Yes I follow the approach from the websockets module as it was proposed in:
- https://github.com/grafana/k6/issues/746#issuecomment-414150215

Reply:
@mstoykov the format is well described here:
https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format

Pretty simple